### ISSUE TITLE
The 'cffi>=1.4.1' distribution was not found and is required by cryptography

### ISSUE THREAD

**Issue by gigabytecoin** [certbot/certbot#3200]
Just installed certbot from the debian Jessie 8.5 backports repository and ran:

`sudo certbot certonly --webroot -w /var/www/example_com/subdomains/www -d example.com -d www.example.com`

Which resulted in:

```
Traceback (most recent call last):
  File "/usr/bin/certbot", line 5, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3138, in <module>
    @_call_aside
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3124, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3151, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 661, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 962, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 849, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'cffi>=1.4.1' distribution was not found and is required by cryptography
```
I successfully created a Let's Encrypt SSL certificate just yesterday following the same process... what happened between now and then?

**Comment by bmw** [comment]
cc @hlieberman

**Comment by spieiga** [comment]
I'm also seeing dependency issues .. when I do an apt-get upgrade I see `python-cryptography` is kept back.

**Comment by zertrin** [comment]
Today I also saw an upgrade for `python-cryptography` from `1.1.1-1~bpo8+1` to `1.3.4-1~bpo8+1` which creates some dependency conflicts that must be resolved manually. 

The first solution proposed wants to uninstall certbot which is obviously not what we want.

By going through the solutions proposed, I finally found one that seems correct, and which leads to promote two packages versions from stable to the jessie-backports:

```
python-pkg-resources:amd64 (5.5.1-1 --> 18.8-1~bpo8+1),
python-setuptools:amd64 (5.5.1-1 --> 18.8-1~bpo8+1)
```

because `python-cryptography 1.3.4-1~bpo8+1` depends on `python-setuptools >= 11.3` which is not available in stable.

After choosing this resolution I thought everything would go well, but still had the error 

```
pkg_resources.DistributionNotFound: The 'cffi>=1.4.1' distribution was not found and is required by cryptography
```

Finally I got it working by installing manually `python-cffi:amd64 (1.4.2-2~bpo8+1)`.

But this is weird because `python-cryptography 1.3.4-1~bpo8+1` depends on `python-cffi-backend-api-9729 (1.4.2-2~bpo8+1)` which I hoped would suffice.

In any case this issue seems to come from the packaging of `python-cryptography` and not certbot, since certbot only depends on `python-cryptography >= 0.7`

This should maybe go to `python-cryptography` upstream packaging team.

**Comment by gigabytecoin** [comment]
> This should maybe go to `python-cryptography` upstream packaging team.

So is `certbot` on debian Jessie essentially broken until the `python-cryptography` team can implement a fix?

**Comment by hlieberman** [comment]
This shouldn't be necessary at all, and I thought this was fixed in the latest version. Let me dig a bit.

**Comment by hlieberman** [comment]
Hello everyone.  This should be fixed now, thanks to the python-cryptography maintainer.  An upgrade should bring in python-cryptography version 1.3.4-1~bpo8+2, which has the fix.  If not, you can use `apt install -t jessie-backports python-cffi` to manually resolve the issue.

**Comment by spieiga** [comment]
I had to do `apt-get install -t jessie-backports python-cryptography` and it seems to be working now.  Thanks @hlieberman !

**Comment by gigabytecoin** [comment]
Perhaps I'm not understanding this line from the [official Debian Backports page](https://backports.debian.org/) properly:

> It is therefore recommended to only select single backported packages that fit your needs, and not use all available backports.

But it seems to insinuate to me at least that using more than one backported package is against Debian stable standards?

Or do they mean simply not to use every single backported package on your system, and that picking a few here or there is OK?

Either way, huge props to the `certbot` and `python-cryptography` teams for getting a fix out in 24 hours!

**Comment by spieiga** [comment]
@gigabytecoin the real question is, when will we get off backports?

**Comment by gigabytecoin** [comment]
@spieiga I'm guessing we will be ready by the time Debian 9 is released in about one year's time.

I vote that this issue be closed.

I just ran:

`sudo aptitude update; sudo aptitude full-upgrade`

And only the `python-cryptography` package was updated.

I was immediately able to install working certificates via certbot after that.

Many thanks for fixing this issue to all involved.

**Comment by magandrez** [comment]
Hi,

I have this very same problem in my machine but apt-get update python-cryptography did not help. Here is what showpkg says:
```

Package: python-cryptography
Versions: 
1.3.4-1~bpo8+2 (/var/lib/apt/lists/ftp.debian.org_debian_dists_jessie-backports_main_binary-amd64_Packages) (/var/lib/dpkg/status)
 Description Language: 
                 File: /var/lib/apt/lists/debian.simnet.is_debian_dists_jessie_main_binary-amd64_Packages
                  MD5: 2e4e904eba9febc1805441f8f837db86
 Description Language: en
                 File: /var/lib/apt/lists/debian.simnet.is_debian_dists_jessie_main_i18n_Translation-en
                  MD5: 2e4e904eba9febc1805441f8f837db86

0.6.1-1 (/var/lib/apt/lists/debian.simnet.is_debian_dists_jessie_main_binary-amd64_Packages)
 Description Language: 
                 File: /var/lib/apt/lists/debian.simnet.is_debian_dists_jessie_main_binary-amd64_Packages
                  MD5: 2e4e904eba9febc1805441f8f837db86
 Description Language: en
                 File: /var/lib/apt/lists/debian.simnet.is_debian_dists_jessie_main_i18n_Translation-en
                  MD5: 2e4e904eba9febc1805441f8f837db86


Reverse Depends: 
  python-cffi-backend,python-cryptography 0.8.2-4~
  python-certbot,python-cryptography 0.7
  python-castellan,python-cryptography 1.0
  python-acme,python-cryptography 0.8
  python-openssl,python-cryptography 1.3
  python-nova,python-cryptography 1.0
  python-magnum,python-cryptography 1.0
  python-keystone,python-cryptography 1.0
  python-heat,python-cryptography 1.0
  python-glance,python-cryptography 1.0
  python-barbican,python-cryptography 1.0
  python-openssl,python-cryptography
Dependencies: 
1.3.4-1~bpo8+2 - python (3 2.8) python (2 2.7~) python-cffi-backend-api-9729 (0 (null)) python-enum34 (0 (null)) python-idna (2 2.0) python-ipaddress (0 (null)) python-pyasn1 (2 0.1.8) python-setuptools (2 11.3) python-six (2 1.4.1) python:any (2 2.7.5-5~) libc6 (2 2.14) libssl1.0.0 (2 1.0.1) python-cryptography-doc (0 (null)) python-cryptography-vectors (0 (null)) 
0.6.1-1 - python:any (2 2.7.5-5~) python-cffi (0 (null)) python (2 2.7~) python (3 2.8) python-pkg-resources (0 (null)) python-six (0 (null)) libc6 (2 2.4) libssl1.0.0 (2 1.0.1) python-cryptography-doc (0 (null)) python-cryptography-vectors (0 (null)) 
Provides: 
1.3.4-1~bpo8+2 - 
0.6.1-1 - 
Reverse Provides: 
```
When running `certbot --apache` I get the same error as described above:
```

Traceback (most recent call last):
  File "/usr/local/bin/certbot", line 5, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2927, in <module>
    @_call_aside
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2913, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2940, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 637, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 650, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 829, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'acme==0.9.0.dev0' distribution was not found and is required by certbot
```
My Debian is fully updated otherwise. Any ideas on why this is happening?

**Comment by hlieberman** [comment]
Hello Manuel,

You don't appear to be using the official packages; there's no
0.9.0.dev0 version that's ever been issued to jessie-backports (or any
other release, for that matter), nor do the official packages install
into /usr/local/bin.

Sincerely,
-- 
Harlan Lieberman-Berg
~hlieberman

**Comment by magandrez** [comment]
Hi @hlieberman ,

Hmmm, I followed the instructions lay down in the documentation (https://certbot.eff.org/all-instructions/#debian-8-jessie-apache). I also added backports to the sources.list.

Any idea why certbot installs in /usr/local/bin and where does that version of the package comes from? I have removed python-certbot-apache and run autoremove after removal and tried the instructions again with the same result.
